include/envoy/access_log/access_log.h

```diff
@@ -74,8 +74,24 @@ using FilterPtr = std::unique_ptr<Filter>;
  */
 class Instance {
 public:
+  Instance(AccessLog::FilterPtr filter) : filter_(std::move(filter)) {}
+
   virtual ~Instance() = default;
 
+  /**
+   * Log a completed request if the underlying AccessLog `filter_` allows it.
+   */
+  virtual void maybeLog(const Http::HeaderMap* request_headers,
+                        const Http::HeaderMap* response_headers,
+                        const Http::HeaderMap* response_trailers,
+                        const StreamInfo::StreamInfo& stream_info) {
+    if (filter_ &&
```

Comment:
The types under include (see https://github.com/envoyproxy/envoy/blob/master/REPO_LAYOUT.md) represent Envoy's public API, and as such I don't think this class should have a constructor, methods that aren't pure virtual, or member fields.

I think there are a couple of ways to do what you want, but my preference is to add a base class (probably in `source/extensions/access_loggers/common`). The base class would be the implementation you have here with different naming. The advantage of making the change in this way is that the `AccessLog::Interface` doesn't change, so any custom loggers maintained outside of Envoy will continue to conform to the API (and can convert to the new base class at their leisure).

Reply:
Done! I added it as AccessLog::InstanceImpl in `source/common/access_log/access_log_impl.cc`. Is that OK?

I did preserve the interface change, switching `log()` to `maybeLog()`, as that felt like a clear interface change regardless of whether one uses the new base class or not.